### DISCUSSION TITLE
Routing access failure after server Nginx deployment

### DISCUSSION THREAD

**Discussion by Charles7c** [vuejs/vitepress#2837] [Q&A] [answered]
### Describe the bug

Recently, I deployed a document site on my server through Nginx and found that if I directly access the specified route, it will be forwarded to the homepage, but the routing address is the one that was accessed. At this point, clicking on other navigation options on the page would be confusing. And I have deployed a site on GitHub without experiencing this issue. I suspect it is a Nginx configuration issue. Can you help me take a look at the problem?

### Reproduction

1.pnpm build
2.upload the content from dist to Nginx server
3.accessing Sites
![新标签页 - Google Chrome 2023-08-17 16-08-40](https://github.com/vuejs/vitepress/assets/25446948/87186df7-5d8d-4585-a8e0-f38caf354d37)

### Expected behavior

Expect normal routing access.

### System Info

```sh
System:
    OS: Windows 10 10.0.19045
    CPU: (24) x64 13th Gen Intel(R) Core(TM) i7-13700K
    Memory: 21.30 GB / 31.75 GB
  Binaries:
    Node: 18.16.0 - D:\Develop\nodejs\node.EXE
    npm: 9.5.1 - D:\Develop\nodejs\npm.CMD
    pnpm: 8.6.12 - ~\AppData\Roaming\npm\pnpm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (115.0.1901.203)
    Internet Explorer: 11.0.19041.1566
```

### Additional context

N/A

### Validations

- [X] Check if you're on the [latest VitePress version](https://github.com/vuejs/vitepress/releases/latest).
- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.dev).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

**Comment by kiaking** [comment]
Please use [Discord](https://discord.com/invite/HBherRA) or [GitHub Discussions](https://github.com/vuejs/vitepress/discussions) for a question 👍 

And when you do, don't forget to paste your Nginx config so that people can see what's wrong with it 👀

**Comment by Charles7c** [comment]
I have preliminarily resolved this issue by configuring cleanUrls to true in vitepress, the article directory structure of my project does not meet the corresponding requirements. Now I have removed this configuration, which means changing from `/foo` to `/foo.html`.

**Comment by Charles7c** [comment]
Do you have any other good solutions or configurations? Welcome to share them?

**Reply by brc-dd** [reply]
I haven't used nginx much. Someone from the community might have insights. If there is a good enough configuration, we can add it to the deploy docs too.

**Comment by Lazydd** [comment, marked as answer]
nginx加这个，匹配的路径`/docs/`改成你的
`location /docs/ {
        index index.html index.htm;
        try_files $uri $uri.html $uri/ /index.html;
 }`

**Reply by Charles7c** [reply]
@brc-dd Hi, this is a no-brainer solution.

**Reply by brc-dd** [reply]
Have you verified if this works fine both with and without `cleanUrls`? Also, does 404 needs any special handling? If it's working, ping me again, I'll update the docs.

**Reply by Charles7c** [reply]
Yes, I have tested both with `cleanUrls` set to true and without it, and the configuration resolves my issue. If the resource indeed does not exist, it will redirect to the homepage. In theory, this configuration is not problematic. In the provided `try_files` directive, the resource lookup order is specified, and if it cannot be found, it returns the content of /index.html. Therefore, a special treatment may be needed for `404` situations.

![image](https://github.com/vuejs/vitepress/assets/25446948/28dbe933-abb6-49b8-93ce-cc5f4abc0db0)

**Reply by Charles7c** [reply]
I temporarily disabled the configuration of `cleanUrls` for now. :)

**Reply by qweasdzxcpkh** [reply]
hi, i just add an other trick method to resolve this, in my case i have custom component at home page:
![image](https://github.com/vuejs/vitepress/assets/26856368/45dd8966-77f2-4e2a-9165-e7041dbb2aa5)
that component has multi link writen by `<a>`, it also have these problem, but when i add `target="self"` attribute for the link, it solve, because of this:
https://stackoverflow.com/a/28802951/24381611
again, it is a trick method. if you can control your nginx config, don't use this


btw, i am curious about docs:preview command have this issue with vitepress@1.2.3, is it any work going on with this? @brc-dd

**Reply by brc-dd** [reply]
docs:preview doesn't have any issue. Please create a separate issue if docs:preview is not working. Regarding nginx, it's fucked up and none of the configurations here perfectly work.

**Reply by qweasdzxcpkh** [reply]
> docs:preview doesn't have any issue. Please create a separate issue if docs:preview is not working. Regarding nginx, it's fucked up and none of the configurations here perfectly work.

thanks, i finally figure it out was the custom component using element-plus's el-popover component, it requires `<client-only>`.

**Comment by meduzen** [comment]
**Edit**: I published a [blog post about this 👇](https://blog.mehdi.cc/articles/vitepress-cleanurls-on-nginx-environment).

Hi! I’m only starting with Vitepress, but here’s how I solved it with Nginx:

1. Enable [`cleanUrls: true`](https://vitepress.dev/guide/routing#generating-clean-url).
2. Use the following Nginx [`location`](https://nginx.org/en/docs/http/ngx_http_core_module.html#location) block:

```nginx
    location / {

        # defines which custom page to serve for 404 status code
        error_page 404 /404.html;

        # if `foo/bar` nor `foo/bar.html` nor `foo/bar/index.html` can’t work, answer with 404
        try_files $uri $uri.html $uri/ =404;
    }
```

<details>
<summary>This has been tested on the following files structure.</summary>

| Markdown | HTML build | Valid path(s) | Invalid path |
|---|---|---|---|
| `/index.md` | `/index.html` | `/` and `/index.html` | - |
| `/notes.md` | `/notes.html` | `/notes` and `/notes.html` | `/notes/` (403, see next paragraph)
| `/notes/my-note.md` | `/notes/my-note.html` | `/notes/my-note`, `/notes/my-note/`, `/notes/my-note.html` | -
| - | `404.html` | - | `/lol`, `/lol/`, `/lol/index.html`, `/lol/whatever.html`

</details>

3. To avoid the 403 on paths with a trailing slash (e.g. `my-blog.com/` or `my-blog.com/notes/`), we can redirect them to their non-trailing slash version using the [`rewrite`](https://nginx.org/en/docs/http/ngx_http_rewrite_module.html#rewrite) directive in the [`server`](https://nginx.org/en/docs/http/ngx_http_core_module.html#server) block:

```nginx
server {
    # many things here…

    # remove trailing slash
    rewrite ^(.+)/+$ $1 permanent;

    location / {

        # defines which custom page to serve for 404 status code
        error_page 404 /404.html;

        # if `foo/bar` nor `foo/bar.html` nor `foor/bar/index.html` can’t work, answer with 404
        try_files $uri $uri.html $uri/ =404;
    }
}
```

Note that the `permanent` keyword do a 301 redirect, which is permanent and might be cached during a very long time by your browser (and also a while longer by search engines). If for some reason you want it to be temporary (302), use the `redirect` keyword.

Well, I think this is gonna be my first article on my blog using Vitepress. 😄  I’ll probably also PR this for the [Vitepress deploy guide](https://vitepress.dev/guide/deploy).

**Reply by brc-dd** [reply]
Just a word of caution, removing trailing slash might break relative url (for links as well as assets). These are resolved by the browser and vitepress has no control over them. This is not an issue with assets statically referenced in markdown or vue, but rather with dynamic things (like logo in theme config, or maybe some data in frontmatter).

**Reply by meduzen** [reply]
If you have an example of this, I’m curious. I don’t see any scenario where a trailing slash is applicable to something else than a URL in the context of Vitepress. Maybe something I still have to understand or discover. 🤔

**Reply by brc-dd** [reply]
Suppose you've something like this in your some page say `foo/index.md` (and you've image at `public/logo.png`

```md
---
layout: home

hero:
  name: VitePress
  text: Vite & Vue powered static site generator.
  tagline: Lorem ipsum...
  image:
    src: '../logo.png'
    alt: VitePress
---
```

here logo.png will break if you remove trailing slash from `/foo/`. you'll need to reference it as `/logo.png` or if you've base, then `/base/logo.png`

**Reply by meduzen** [reply]
Just tested with the following file structure, and I don’t have the issue you mention.

Using Vitepress config with `cleanUrls: true` and the Nginx config from my first answer.

As input:
- `/public/logo.png`
- `/notes/my-title.md`:

```md
---
layout: home

hero:
  name: Test
  text: Let’s test something.
  tagline: Lorem ipsum...
  image:
    src: '../logo.png'
    alt: VitePress
---

We are in `/notes/my-title.md`, and frontmatter `hero.image.src` is `'../logo.png'`.
```

Build:
- `/logo.png`
- `/notes/my-title.html`

Try it by yourself on these URLs (I’ll unpublish them once the conversation is resolved):
- `https://blog.mehdi.cc/notes/my-title`
- `https://blog.mehdi.cc/notes/my-title/` (which removes the trailing slash and redirects to `https://blog.mehdi.cc/notes/my-title`)
- `https://blog.mehdi.cc/notes/my-title.html`

You can see the hero picture is `<img class="VPImage image-src" src="../logo.png" alt="VitePress" data-v-8426fc1a="">`, which does not impact its loading as trailing URLs are removed.

**Reply by brc-dd** [reply]
should happen only with xxx/index.md  kind of files. and I don't think you'll be able to remove trailing slash from those files (vitepress probably auto adds it even if you have removed it).

**Reply by meduzen** [reply]
Okay, got it! Indeed, it creates an infinite redirection loop between `/path/` and `/path`.  Only `/test/index.html` is working normally. (I saw you brought many routing feature, I have to look up why Vitepress forces the slash).

So I see 3 paths:
1. Use `/path.md`over `/path/index.md` as it doesn’t break leading slash, which is what I’m doing for now.
2. Don’t remove the trailing slash, but it exposes to a 403.
3. Always add a trailing slash, which at first attempt seems to be more complicated with Nginx.

Thanks for your time, by the way! I still have to experiment with dynamic stuff (e.g. building an index of `/notes/*`).
